src/bootstrap/config.rs

```diff
@@ -798,7 +798,20 @@ impl Config {
         config.llvm_profile_use = flags.llvm_profile_use;
         config.llvm_profile_generate = flags.llvm_profile_generate;
 
-        let stage0_json = t!(std::fs::read(&config.src.join("src").join("stage0.json")));
+        let stage0_json = std::fs::read(&config.src.join("src").join("stage0.json"))
+            .unwrap_or_else(|_| {
+                let base_path = output(Command::new("git").arg("rev-parse").arg("--show-toplevel"))
+                    .trim()
+                    .to_owned();
+
+                config.src = PathBuf::from(base_path);
+                config.initial_rustc =
+                    config.out.join(config.build.triple).join("stage0/bin/rustc");
+                config.initial_cargo =
+                    config.out.join(config.build.triple).join("stage0/bin/cargo");
+                t!(std::fs::read(&config.src.join("src").join("stage0.json")))
+            });
```

Comment:
This looks pretty suspicious to me -- if we need to remap the base path via git rev-parse, that's probably something that should be done when discovering config.src initially, not when failing to read stage0.json.

We should also most likely hook the discovery mechanism to not be opportunistic (on failing to find stage0.json), but rather based on whether we're building for distribution (i.e., in CI), in which case we can be reliant on git and not try to to use the env-variables set by Cargo for these things.

Reply:
> We should also most likely hook the discovery mechanism to not be opportunistic (on failing to find stage0.json), but rather based on whether we're building for distribution (i.e., in CI), in which case we can be reliant on git and not try to to use the env-variables set by Cargo for these things.

@bjorn3 objected to that on zulip:
> Both options are incorrect for CI distributed rustbuild when you are invoking it from another directory than the source directory.

but I agree just using git is more reliable and simpler.

Reply:
I think the future option is probably that we embed a symlink (or, well, text file, hard link, whatever) next to bootstrap pointing to the real source (generated by whatever tool is downloading bootstrap), but for now I'm OK sticking to just using git for this since we don't actually have that tool.

But I do think this should be refactored to be in the right place.